### DISCUSSION TITLE
Best Practices for Updating GraphRAG Index with Frequently Changing Data Sources?

### DISCUSSION THREAD

**Discussion by owquresh** [microsoft/graphrag#1313] [Q&A]
I am currently working on a project where we utilize GraphRAG for query-focused summarization over documents stored in Google Drive. The data in Google Drive changes quite frequently, with documents being added, updated, or deleted on a regular basis.

I’m looking for guidance on the best practices for updating the graph-based index in GraphRAG to keep it in sync with these changes. Specifically, I’d appreciate your advice on:

	1.	Incremental Updates: How can we efficiently update the graph when documents are modified or new documents are added? Should we update only the affected parts of the graph, or is a full reindexing recommended?
	2.	Handling Deletions: What’s the recommended approach for removing entities and relationships from the graph when documents are deleted or content is removed?
	3.	Change Detection: Are there built-in methods in GraphRAG for detecting changes in the dataset (e.g., based on modified timestamps or content hashing), or should we handle this externally?
	4.	Optimizing for Frequent Changes: Given that the data changes quite frequently, are there strategies or configurations within GraphRAG that would help optimize the updating process to avoid unnecessary computations or reprocessing?
	5.	Real-Time or Near Real-Time Indexing: If we need to keep the graph as up-to-date as possible, what’s the best way to approach near real-time updates? Are there considerations we should be aware of when attempting to keep the index in sync with the data in near real-time?

**Comment by AxelSjoberg** [comment]
@owquresh, have you attained any further insights here? We have also been thinking about these exact questions in our team.

Probably already seen this, but in version 1.0 there is an [update command](https://www.microsoft.com/en-us/research/blog/moving-to-graphrag-1-0-streamlining-ergonomics-for-developers-and-users/) (have not tried it myself yet), that should help with point 1 and 3. You can read more about it in this blogpost [here](https://www.microsoft.com/en-us/research/blog/moving-to-graphrag-1-0-streamlining-ergonomics-for-developers-and-users/).

We've been awaiting the release for the LazyGraphRAG implementation also, which I would assume has an update command as well (which hopefully is efficient). If this is the case - the lower indexation cost and flexible document ingestion would help a lot to build more practical applications and address the issues you've raised.
